### PR TITLE
Begin Profiling Rework to move towards Modularity

### PR DESCRIPTION
**Pull request by maiadegraaf**
_Had to open a new PR but it is the same as #10935._

This PR introduces a new pragma `custom_profiling_settings` which accepts a path to a JSON file.  The file can be used to turn on or off profiling settings when the profiling mode is set to JSON (`PRAGMA enable_profiling = 'json';`).  If the pragma is not set, all profiling information will be included.

Currently, the settings are as follows:

- **CPU Timing**: The total CPU time spent by that operator and its children.
- **Operator Cardinality**: The number of rows returned by that operator
- **Operator Timing**: The time spent by that operator
- **Extra Info**: Any other information collected during the operation

An example of a custom profiling settings JSON file:
```sql
PRAGMA custom_profiling_settings='/path/to/settings.json'
```
```json
{
  "cpu_time":"true",
  "operator_cardinality":"true",
}
```

----
The plan is to expand this feature to be able to include more settings in the future, as well as a new function in the C API that returns an object with the profiling information.

### CODE REVIEW

File: src/include/duckdb/main/profiling_info.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+enum class MetricsType : uint8_t { CPU_TIME, EXTRA_INFO, OPERATOR_CARDINALITY, OPERATOR_TIMING };
+
+struct MetricsTypeHashFunction {
+	uint64_t operator()(const MetricsType &index) const {
+		return std::hash<uint8_t>()(static_cast<uint8_t>(index));
+	}
+};
+
+typedef unordered_set<MetricsType, MetricsTypeHashFunction> SettingsSet;
```
Comment (taniabogatsch) [inline src/include/duckdb/main/profiling_info.hpp:29]: Nit: can we rename this to `ProfilerSettings`?

File: src/include/duckdb/main/profiling_info.hpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	idx_t operator_cardinality;
+	double operator_timing;
+
+	Metrics() : cpu_time(0), operator_cardinality(0), operator_timing(0) {
+	}
+};
+
+class ProfilingInfo {
+public:
+	// map of metrics with their values; only enabled metrics are present in the map
```
Comment (taniabogatsch) [inline src/include/duckdb/main/profiling_info.hpp:55]: Nit: no longer a `map`.

File: src/include/duckdb/main/profiling_info.hpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+public:
+	ProfilingInfo() = default;
+	ProfilingInfo(SettingsSet &n_settings) : settings(n_settings) {
+	}
+	ProfilingInfo(ProfilingInfo &) = default;
+	ProfilingInfo &operator=(ProfilingInfo const &) = default;
+
+public:
+	// set the metrics map
```
Comment (taniabogatsch) [inline src/include/duckdb/main/profiling_info.hpp:67]: Nit: no longer a `map`. Parameter can be `const`?

File: src/include/duckdb/main/profiling_info.hpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	void SetSettings(SettingsSet &n_settings);
+	// get the metrics map
```
Comment (taniabogatsch) [inline src/include/duckdb/main/profiling_info.hpp:69]: Nit: no longer a `map`. Should this return a `const` reference?

File: src/include/duckdb/main/settings.hpp [hunk 5]
```diff
@@ -267,6 +267,16 @@ struct EnableProfilingSetting {
 	static Value GetSetting(ClientContext &context);
 };
 
+struct CustomProfilingSettings {
+	static constexpr const char *Name = "custom_profiling_settings";
+	static constexpr const char *Description =
+	    "Accepts a path to a JSON file which is used to customize the profiling statistics";
```
Comment (taniabogatsch) [inline src/include/duckdb/main/settings.hpp:273]: `"Accepts a path to a JSON file enabling custom metrics."

File: src/include/duckdb/main/query_profiler.hpp [hunk 6]
```diff
@@ -80,46 +82,61 @@ struct ExpressionExecutorInfo {
...
 
-	double time = 0;
-	idx_t elements = 0;
+	double time;
+	idx_t elements;
 	string name;
 	//! A vector of Expression Executor Info
 	vector<unique_ptr<ExpressionExecutorInfo>> executors_info;
+
+	void addTime(double n_time) {
```
Comment (taniabogatsch) [inline src/include/duckdb/main/query_profiler.hpp:98]: Uppercase.

File: src/include/duckdb/main/query_profiler.hpp [hunk 7]
```diff
@@ -80,46 +82,61 @@ struct ExpressionExecutorInfo {
...
+		this->time += n_time;
+	}
+
+	void addElements(idx_t n_elements) {
```
Comment (taniabogatsch) [inline src/include/duckdb/main/query_profiler.hpp:102]: Uppercase.

File: src/include/duckdb/main/query_profiler.hpp [hunk 8]
```diff
@@ -128,22 +145,29 @@ class QueryProfiler {
 	DUCKDB_API QueryProfiler(ClientContext &context);
 
 public:
+	// Recursive tree that mirrors the operator tree
 	struct TreeNode {
 		PhysicalOperatorType type;
 		string name;
-		string extra_info;
-		OperatorInformation info;
+		ProfilingInfo settings;
```
Comment (taniabogatsch) [inline src/include/duckdb/main/query_profiler.hpp:152]: Nit: this includes both the settings and the metrics now.

File: src/include/duckdb/main/query_profiler.hpp [hunk 9]
```diff
@@ -128,22 +145,29 @@ class QueryProfiler {
...
+	};
+
 	// Propagate save_location, enabled, detailed_enabled and automatic_print_format.
 	void Propagate(QueryProfiler &qp);
 
 	using TreeMap = reference_map_t<const PhysicalOperator, reference<TreeNode>>;
 
 private:
-	unique_ptr<TreeNode> CreateTree(const PhysicalOperator &root, idx_t depth = 0);
+	unique_ptr<QueryProfiler::TreeNode> CreateTree(const PhysicalOperator &root, SettingsSet settings, idx_t depth = 0);
```
Comment (taniabogatsch) [inline src/include/duckdb/main/query_profiler.hpp:170]: This cannot be `<TreeNode>` any longer?

File: src/main/query_profiler.cpp [hunk 10]
```diff
@@ -126,33 +126,59 @@ void QueryProfiler::StartExplainAnalyze() {
...
-	if (root) {
+	if (root && root->settings.Enabled(MetricsType::OPERATOR_CARDINALITY)) {
 		Finalize(*root);
 	}
 	this->running = false;
 	// print or output the query profiling after termination
 	// EXPLAIN ANALYSE should not be outputted by the profiler
 	if (IsEnabled() && !is_explain_analyze) {
-		string query_info = ToString();
+		// add an outer node for the query
```
Comment (taniabogatsch) [inline src/main/query_profiler.cpp:153]: Nit: no longer a node.

**Review by taniabogatsch — APPROVED** [review]
Looks neat now! I only had a few nits. :)

**Comment by taniabogatsch** [comment]
Waiting on https://github.com/duckdb/duckdb/pull/11868.
@maiadegraaf, please ping me once this is ready for review!

**Comment by Maxxen** [comment]
https://github.com/duckdb/duckdb/pull/11868 has been superseded by https://github.com/duckdb/duckdb/pull/11998 so this should be unblocked now!

**Comment by taniabogatsch** [comment]
Still waiting on merging `main` into `feature` to unlock the new JSON features.

**Review by taniabogatsch — APPROVED** [review]
The CI failure seems unrelated.

**Comment by maiadegraaf** [comment]
> The CI failure seems unrelated.

Its because I changed the name of certain proponents in the profiling, I updated the script so should work from now on. Just fails in the CI on this PR.

**Pull request merged**